src/services/types.ts

```diff
@@ -238,8 +238,8 @@ namespace ts {
 
         getCompletionsAtPosition(fileName: string, position: number): CompletionInfo;
         // "options" is optional only for backwards-compatibility
-        getCompletionEntryDetails(fileName: string, position: number, entryName: string, options?: FormatCodeOptions | FormatCodeSettings): CompletionEntryDetails;
-        getCompletionEntrySymbol(fileName: string, position: number, entryName: string): Symbol;
+        getCompletionEntryDetails(fileName: string, position: number, id: string | CompletionEntryIdentifier, options?: FormatCodeOptions | FormatCodeSettings): CompletionEntryDetails;

```

Comment:
this is going to be a breaking change for plugin authors. pinging @chuckjaz and @angelozerr

Reply:
an alternative is to make the source as a third optional parameter.

Reply:
To me this change seems unnecessary as it assumes that the key to the entry is now the `name` + `source` where `name` should be sufficient. Currently, if I am adding something to the completion list I need to ensure that my `name`s are unique already so the call to `getCompletionEntryDetails` just needs the `name`.

Reply:
> `name` should be sufficient

It isn't -- different completions (with different associated code actions) could share the same name.

Reply:
It is unfortunate already merge this.

Adding an optional parameter was arguably a worse breaking change than changing the type of one of the parameters. Both break clients of the API but adding an API will cause all existing plugins to appear to break TypeScript as they will silently miss passing the additional parameter as well as the compiler will not point out that this happened. Changing the type would still be a breaking change but, at least, existing implementation that wrap the language service would probably just pass the parameter unmolested to the wrapped version of the language service and it would work and recompiling the plugin would point out that something changed. In other words, if this change had not been brought to my attention I probably would have missed it entirely, publishing a broken language plugin.

The fundamental problem here is that `name` is treated as `insertText` and they should be different. This would allow plugins to be able to add duplicate `insertText` without repeating the `name`.  `name` should have just been an arbitrary string that was guaranteed to be unique for each entry. Adding `source` as a documentation seems reasonable.

Reply:
@chuckjaz chatted with @andy-ms and he has a proposal of making the new parameter of type `source: string | undefined`. this way it is required, and plugin authors would get a compiler error.

We would rather have it work at runtime for non-module entries than fail all the time. making the name `string | {name: string,source: string}` have a higher chance of breaking at runtime. 

As for the name being different from the insertText, i think this should have been the right design, but i am not sure we can do this now without a big breaking change.

Reply:
@mhegazy This is a breaking change for the plugin. If someone loads a any version of the `@angular/language-service` in a version of TypeScript that has this change it will break. The current plugins have no way to know that this happened and will break any call to `getCompletionEntryDetails` or `getCompletionEntrySymbol` that passes the additional parameter (rather a subtle break that will either go unnoticed or incorrectly diagnosed for quite a while).

New version of the plugin will work with older version of TypeScript (as passing an extra parameter is ignored) but old version of the plugin will break with new version of TypeScript.

It would be nice to adjust the plugin spec with something such that I can declare the I was written knowing this and have the plugin system refuse to load a plugin if it doesn't contain the marker and refuse to use the the plugin if it either doesn't contain the marker as well as emitting an appropriate diagnostic. Maybe something like `export SUPPORTS_PLUGIN_VERSION_2  = true;`. This way the modified TypeScript will not load old version of the Angular plugin.

As the current set of plugins is small maybe we can add this with little impact.

Reply:
The plugin can check the ts.version and you can decide to throw if it is higher than 2.6 for instance.

Reply:
we also have `ts.servicesVersion` that we can change, and then you can do the check. I would say this is cleaner than ts.version, since the two change independently.

if this works for you, we can rev tat to `0.6` (currently `0.5`).

Reply:
The issue is in plugins that have already shipped not in new plugins. Checking the version doesn't help older versions. The version is fine for creating a shim for older versions of the API so a plugin can support older versions of TypeScript but a breaking change a `@anglular/language-service@4.4.5` does not contain the code, couldn't have known that 2.6 would break it to even know to refuse to load. Only 2.6 knows it broke the API. And only it can decide if the an older version of a plugin that doesn't know about its change will work.

Reply:
I understand. i am trying to solve it for future versions. what i am proposing is use the API version. we will increment the API version every time we make a breaking change (that is `ts.servicesVersion`) independently from the version of the language/toolset. the plugin can check at load time if the API version is the version it expects, and if not it can chose to fail, or just issue a warning.
would that work?

Reply:
The older, already shipped, version of the language service do not check this. This would guard everything from `5.0.1+` and a LTS version both of which would then check for 0.6 or 0.5 which they both could support.